llvm/lib/CodeGen/CodeGenPrepare.cpp

```diff
@@ -2114,19 +2114,30 @@ bool CodeGenPrepare::optimizeURem(Instruction *Rem) {
 /// Some targets have better codegen for `ctpop(X) u< 2` than `ctpop(X) == 1`.
 /// This function converts `ctpop(X) ==/!= 1` into `ctpop(X) u</u> 2/1` if the
 /// result cannot be zero.
-static bool adjustIsPower2Test(CmpInst *Cmp, const DataLayout &DL) {
+static bool adjustIsPower2Test(CmpInst *Cmp, const TargetLowering &TLI,
+                               const TargetTransformInfo &TTI,
+                               const DataLayout &DL) {
   ICmpInst::Predicate Pred;
   if (!match(Cmp, m_ICmp(Pred, m_Intrinsic<Intrinsic::ctpop>(), m_One())))
     return false;
   if (!ICmpInst::isEquality(Pred))
     return false;
   auto *II = cast<IntrinsicInst>(Cmp->getOperand(0));
+
+  // Check if it is profitable for the target
+  ICmpInst::Predicate NewPred =
+      Pred == ICmpInst::ICMP_EQ ? ICmpInst::ICMP_ULT : ICmpInst::ICMP_UGT;
+  if (TLI.isCtpopFast(TLI.getValueType(DL, II->getType())) &&
```

Comment:
Doesn't depend on ctpop speed? This is only whether the predicate and/or constant is cheaper?

Reply:
I'd prefer it if this cost model check were dropped again.